src/judge.rs

```diff
@@ -104,8 +95,8 @@ impl Judge {
         Outcome::AttackerWins(weak_defenders)
     }
 
-    fn valid(&self, word: &String) -> bool {
-        self.dictionary.contains(word)
+    fn valid<S: AsRef<str>>(&self, word: S) -> bool {
+        self.dictionary.contains(&word.as_ref().to_lowercase())
     }
```

Comment:
Tip: Take a generic `S: AsRef<str>` as much as possible. 

Here we're saying that word is a generic type `S`, which has to implement `AsRef<str>`. This means we have to wrangle string types less, as this function will accept any owned or non-owned string (they can all auto ref into a `&str`) — see the tests file below

Reply:
(Maybe some people would take issue with "as much as possible" — if you're definitely only ever passing one type in — but there's no real downside to the generic function if you're only ever passing one type in so 🤷‍♂️ I tend to write these)

Reply:
OK nice, that's kinda what I originally thought I was doing when I used `&str`, because it felt like any kind of string should be able to be a non mutable borrow of a str. But I guess when you have a `&str` parameter you actually need a concrete `&str` argument, not just something that could become a `&str`.

I guess super generic functions can be less readable if they're only going to be used for one type. But I think `AsRef<str>` is probably clear enough that we can use it "as much as possible."